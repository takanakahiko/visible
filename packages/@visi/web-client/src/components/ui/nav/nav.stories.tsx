import React from 'react';

import { Nav } from '.';

export default {
  title: 'Nav',
  component: Nav,
};

export const Normal = () => (
  <Nav
    items={[
      { href: '/foo', as: '/bar', children: 'Click me' },
      { href: '/foo', as: '/bar', children: 'Click me' },
    ]}
  />
);
