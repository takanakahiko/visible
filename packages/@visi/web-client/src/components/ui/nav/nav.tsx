import classNames from 'classnames';
import Link from 'next/link';
import { useRouter } from 'next/router';
import React, { createContext, useContext } from 'react';

export type NavVariant = 'normal' | 'invert';
export type NavDirection = 'horizontal' | 'vertical';
export type NavSpace = 'lg' | 'md' | 'sm';

const NavContext = createContext<{
  variant?: NavVariant;
  direction?: NavDirection;
}>({});

const useVariant = () => {
  return useContext(NavContext).variant;
};

export interface NavItem {
  href: string;
  as?: string;
  children: React.ReactNode;
  icon?: React.ReactNode;
}

const Item = (props: NavItem) => {
  const { href, as: _as, children, icon } = props;

  const router = useRouter();
  const variant = useVariant();

  const active = router.pathname === href;
  const isExternal = _as == null;

  const Wrapper = isExternal
    ? React.Fragment
    : ({ children }: { children: React.ReactNode }) => (
        <Link href={href} as={_as}>
          {children}
        </Link>
      );

  return (
    <Wrapper>
      <a
        className={classNames(
          'py-2',
          'px-3',
          'rounded-md',
          'text-sm',
          'font-semibold',
          'space-x-2',
          'transition-colors',
          'duration-75',
          variant === 'normal' &&
            (active
              ? ['text-gray-600', 'bg-gray-300']
              : ['text-gray-600', 'hover:bg-gray-300']),
          variant === 'invert' &&
            (active
              ? ['text-white', 'bg-primary-600']
              : ['text-white', 'hover:bg-primary-600']),
        )}
        {...(isExternal ? { href, target: '_blank', rel: 'noopener' } : {})}
      >
        {icon && <span>{icon}</span>}
        <span>{children}</span>
      </a>
    </Wrapper>
  );
};

export interface NavProps {
  items: NavItem[];
  variant: NavVariant;
  direction: NavDirection;
  space: NavSpace;
}

export const Nav = (props: NavProps) => {
  const { items, variant, direction, space } = props;

  const infix = {
    lg: '8',
    md: '4',
    sm: '2',
  }[space];

  return (
    <NavContext.Provider value={{ variant, direction }}>
      <nav>
        <ul
          className={classNames(
            'flex',
            direction === 'horizontal'
              ? ['flex-row', `space-x-${infix}`]
              : ['flex-col', `space-y-${infix}`],
          )}
        >
          {items.map((item, i) => (
            <li key={`${item.href}-${i}`}>
              <Item {...item} />
            </li>
          ))}
        </ul>
      </nav>
    </NavContext.Provider>
  );
};

Nav.defaultProps = {
  space: 'md',
  variant: 'normal',
  direction: 'horizontal',
};
