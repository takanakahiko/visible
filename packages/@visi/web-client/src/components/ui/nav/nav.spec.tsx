import { render } from '@testing-library/react';
import React from 'react';

import { Normal } from './nav.stories';

describe('Nav', () => {
  it('matches snapshot', () => {
    const { container } = render(<Normal />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
