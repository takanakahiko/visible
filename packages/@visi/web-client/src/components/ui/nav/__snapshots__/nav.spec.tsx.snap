// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav matches snapshot 1`] = `
<nav>
  <ul
    class="flex flex-row space-x-4"
  >
    <li>
      <a
        class="py-2 px-3 rounded-md text-sm font-semibold space-x-2 transition-colors duration-75 text-gray-600 hover:bg-gray-300"
        href="/bar"
      >
        <span>
          Click me
        </span>
      </a>
    </li>
    <li>
      <a
        class="py-2 px-3 rounded-md text-sm font-semibold space-x-2 transition-colors duration-75 text-gray-600 hover:bg-gray-300"
        href="/bar"
      >
        <span>
          Click me
        </span>
      </a>
    </li>
  </ul>
</nav>
`;
